regression/input/typemap.yaml

```diff
@@ -0,0 +1,75 @@
+copyright:
+- Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+- other Shroud Project Developers.
+- See the top-level COPYRIGHT file for details.
+-
+- "SPDX-License-Identifier: (BSD-3-Clause)"
+-
+
+library: typemap
+cxx_header: typemap.hpp
+
+typemap:
+# IndexType may be 32 or 64 bits depending on USE_64BIT_INDEXTYPE
+- type: IndexType
+  fields:
+    base: integer
+    f_kind: INDEXTYPE
+    f_module_name: typemap_mod
+
+- type: FloatType
+  fields:
+    base: real
+    f_kind: FLOATTYPE
+    f_module_name: typemap_mod
+
+declarations:
+#- decl: IndexType i1
+
+- decl: bool passIndex(IndexType i1, IndexType *i2+intent(out))
+
+- decl: void passFloat(FloatType f1);
+
+######################################################################
+splicer_code:
+  f:
+    module_top: |
+      #if defined(USE_64BIT_INDEXTYPE)
+      integer, parameter :: INDEXTYPE = C_INT64_T
+      #else
+      integer, parameter :: INDEXTYPE = C_INT32_T
+      #endif
+
+      #if defined(USE_64BIT_FLOAT)
+      integer, parameter :: FLOATTYPE = C_DOUBLE
+      #else
+      integer, parameter :: FLOATTYPE = C_FLOAT
+      #endif
+  c:
+    C_declarations: |
+      #if defined(USE_64BIT_INDEXTYPE)
+      typedef int64_t IndexType;
+      #else
+      typedef int32_t IndexType;
+      #endif
+
+      #if defined(USE_64BIT_FLOAT)
+      typedef double FloatType;
+      #else
+      typedef float FloatType;
+      #endif
```

Comment:
Clarification: In the axom use-case, would this just be something like `#include axom/config.hpp`, which has configuration details about using 32 vs. 64 bits?

Reply:
These typedefs are added to a ``.h`` file which is "For C users and C++ implementation".  I tested surrounding the *C_declarations* with ``#ifndef __cplusplus`` since the typedefs are only intended for C users. However, I ran into a compile error since ``wraptypemap.cpp``, the C++ wrapper implementation, includes headers in the wrong order.  It should include ``typemap.hpp``, the equivalent of ``axom/config.hpp``, before ``wraptypemap.h``. However, since that fix will touch every test I'll do it in another PR.

Since this is for C users, the typedefs must be duplicated using C syntax instead of ``using IndexType = std::int64_t;``.